### ISSUE TITLE
2.x - Allow customization of span naming logic for traces

### ISSUE THREAD

**Issue by hyleung** [hyleung/ratpack-zipkin#26]
In v1, there was a `SpanNameProvider` interface that a user could provide an implementation of via the module config.

```
public interface SpanNameProvider {
    String spanName(HttpRequest request);
}
```

e.g.

```
config
.serviceName("ratpack-demo")
.spanNameProvider(new SpanNameProvider() {
        @Override
         public String spanName(final HttpRequest request) {
                      return "foo" + request.getHttpMethod();
         }
})
```

For v2, we should provide something similar.  `SpanNameProvider` still exists but we've removed the ability to configure it. Also it still has a lot of the feel of Brave http (adapters, etc.) - so now might be a good opportunity to redesign that stuff (thinking maybe a `@FunctionalInterface`).

**Comment by codefromthecrypt** [comment]
I currently have a plugin with 3 methods spanName requestTags and
responseTags which default but can be overridden as a unit.

This allows you to use the same plugin to coerce any http library (similar
to brave 3) what is different is that the parsing is lazy and also the
input can have other data.

Ex currently thinking brave.http.HttpRequestContext which wraps calls ex
req.method() would call the underlying. However you could subtype this.. ex
RatpackHttpRequestContext which could possibly also expose a connection
which could aid building more sophisticated tags. It could also expose the
raw request object.

Moreover the ResponseContext could carry an error message or throwable.
This allows someone to write the "error" tag differently across any library
defaulting to something sensible.

Will ping with concrete code which will be soon like today

Im currently jigging this across other libraries and want to make sure it
works.

**Comment by codefromthecrypt** [comment]
made some notes here. After looking both ways I ditched the wrapping
approach which I tried for a few hours today. I think this one is best, but
let me know. Regardless, we can choose to make span name parser an atom or
a part of a composite plugin.

https://github.com/openzipkin/brave/pull/350/files#r113442950

**Comment by hyleung** [comment]
@adriancole sorry, I didn't get a chance to look at this until today - the link above just goes to the PR (maybe because of a rebase?). Anyhow - is the idea that we'd extend `brave.ClientHandler` or `brave.Serverhandler`? Or would it be `brave.http.HttpAdapter`?

**Comment by llinder** [comment]
This can be customized with Brave 4.3.1 by providing a custom HttpParser and override the `spanName` method.  The custom HttpParser can be set through the ServerTracingModule `serverParser` and `clientParser`.

@hyleung do you feel this sufficiently covers the ability to customize the span names?

**Comment by hyleung** [comment]
@llinder similar to #27 - think we can punt it for 2.0.0. Again, wanted to see if it would make sense to have something more RatPack-y.

**Comment by hyleung** [comment]
Implemented by #42
